drivers/opt3001/opt3001.c

```diff
@@ -0,0 +1,189 @@
+{
+    /* Check parameters */
+    assert(dev && params);
+
+    uint16_t reg;
+
+    dev->params = *params;
+
+    /* Test devide id */
+    i2c_acquire(DEV_I2C);
+    if (i2c_read_regs(DEV_I2C, DEV_ADDR, OPT3001_REGS_DEVICE_ID, &reg, 2, 0) < 0) {
+        i2c_release(DEV_I2C);
+        LOG_ERROR("opt3001_init: Error reading device ID!\n");
+        return -OPT3001_ERROR_BUS;
+    }
+
+    reg = htons(reg);
+    if (reg != OPT3001_DID_VALUE) {
+        return -OPT3001_ERROR_DEV;
+    }
+
+    /* Set range number, mode of conversion and conversion time */
+    reg = OPT3001_CONFIG_RN_FSR;
+    reg |= OPT3001_REGS_CONFIG_MOC(OPT3001_CONFIG_M_SHUTDOWN);
+    reg |= OPT3001_REGS_CONFIG_CT(OPT3001_CONVERSION_TIME);
+
+    /* Configure for latched window-style comparison operation */
+    reg |= OPT3001_REGS_CONFIG_L;
+    reg &= ~OPT3001_REGS_CONFIG_POL;
+    reg &= ~OPT3001_REGS_CONFIG_ME;
+    reg &= ~OPT3001_REGS_CONFIG_FC_MASK;
+
+    reg = htons(reg);
+
+    if (i2c_write_regs(DEV_I2C, DEV_ADDR, OPT3001_REGS_CONFIG, &reg, 2, 0) < 0) {
+        i2c_release(DEV_I2C);
+        LOG_ERROR("opt3001_init: Error setting device configuration\n");
+        return -OPT3001_ERROR_BUS;
+    }
+    i2c_release(DEV_I2C);
+
+    return OPT3001_OK;
+}
+
+int opt3001_reset(const opt3001_t *dev)
+{
+    uint16_t reg = OPT3001_CONFIG_RESET;
+    reg |= OPT3001_REGS_CONFIG_CT(OPT3001_CONVERSION_TIME);
+    reg = htons(reg);
+
+    /* Acquire exclusive access to the bus. */
+    i2c_acquire(DEV_I2C);
+    if (i2c_write_regs(DEV_I2C, DEV_ADDR, OPT3001_REGS_CONFIG, &reg, 2, 0) < 0) {
+        i2c_release(DEV_I2C);
+        return -OPT3001_ERROR_BUS;
+    }
+    i2c_release(DEV_I2C);
+    return OPT3001_OK;
+}
+
+int opt3001_set_active(const opt3001_t *dev)
+{
+    uint16_t reg;
+
+    i2c_acquire(DEV_I2C);
+
+    if (i2c_read_regs(DEV_I2C, DEV_ADDR, OPT3001_REGS_CONFIG, &reg, 2, 0) < 0) {
+        i2c_release(DEV_I2C);
+        LOG_ERROR("opt3001_init: Error reading BUS!\n");
+        return -OPT3001_ERROR_BUS;
+    }
+
+    reg = htons(reg);
+    reg |= OPT3001_REGS_CONFIG_MOC(OPT3001_CONFIG_M_CONTINUOUS);
+
+    reg = htons(reg);
+    if (i2c_write_regs(DEV_I2C, DEV_ADDR, OPT3001_REGS_CONFIG, &reg, 2, 0) < 0) {
+        i2c_release(DEV_I2C);
+        LOG_ERROR("opt3001_init: Error setting device configuration\n");
+        return -OPT3001_ERROR_BUS;
+    }
+
+    i2c_release(DEV_I2C);
+    return OPT3001_OK;
+}
+
+int opt3001_read(const opt3001_t *dev, uint16_t *crf, uint16_t *rawl)
+{
+    uint16_t reg;
+
+    i2c_acquire(DEV_I2C);
+
+    if (i2c_read_regs(DEV_I2C, DEV_ADDR, OPT3001_REGS_CONFIG, &reg, 2, 0) < 0) {
+        i2c_release(DEV_I2C);
+        LOG_ERROR("opt3001_init: Error reading BUS!\n");
+        return -OPT3001_ERROR_BUS;
+    }
+
+    *crf = htons(reg) & OPT3001_REGS_CONFIG_CRF;
+
+    if (!(*crf)) {
+        i2c_release(DEV_I2C);
+        LOG_DEBUG("opt3001_read: conversion in progress!\n");
+        return -OPT3001_ERROR;
```

Comment:
Instead of returning error here, why not repeat reading `OPT3001_REGS_CONFIG` in a loop until `OPT3001_REGS_CONFIG_CRF` is set?
You can still call `xtimer_usleep()` in the loop, but only sleep 1ms at a time until you have reached `OPT3001_CONVERSION_TIME`.

Another option instead of polling would be to use the interrupt pin to get an interrupt when conversion is done.